2016/proposals/rai-vivek.md

```diff
@@ -0,0 +1,360 @@
+  id PRIMARY KEY AUTO,
+  software VARCHAR,
+  installed BOOL,
+  version VARCHAR,
+  log VARCHAR
+}
+```
+
+Since we are obtaining the platform information using Python `platform` module
+which is same across all platforms, the same database schema can accommodate all
+systems.
+
+A typical example of the database entry then may look like:
+
+```
+(TABLE machine_info)
+
+id | system  | hostname    | release             | version                                               | machine  | processor | email       | time                      | exit_status
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+1  | ‘Linux’ | ‘ghostbook’ | ‘3.19.0-47-generic’ | ‘#53~14.04.1-Ubuntu SMP Mon Jan 18 16:09:14 UTC 2016’ | ‘x86_64’ | ‘x86_64’  | foo@bar.com | ‘Tue Mar 8 00:52:35 2016’ | 0
+
+(TABLE software_version)
+
+id | software | installed | version | log
+-----------------------------------------
+1  | python   | True      | 3.5.1   | ‘’
+1  | git      | True      | 1.9.1   | ‘’
+1  | hg       | False     | ‘’      | ‘’
+1  | pip      | True      | 7.9     | ‘A newer version of pip is available.’
+```
+
+Complex queries can be performed by `JOINING` the two tables and querying for
+corresponding field of interest.
+
+#### Concerns
+Both tables share the unique identity given to a particular report.  There is
+a possibility that same user may submit a report twice possibly because there
+were some errors or something else causing duplication.
+
+There might be a couple of ways to tackle this:
+
+1. One way to avoid it would be to generate a unique `machine-id` and use
+   that as a primary identifier key. This key will act as foreign key for key
+   for other table and maintain consistency.
+2. Just execute a SELECT statement on table `machine-info` and see if that
+   information is already record. If yes, then update the records in
+   `software_version` table. Essentially, all the attributes of `machine-info`
+   table may constitute a `UNIQUE_KEY` constraint.
+
+### User interface
+
+* An easily navigable HTML report of diagnostic data.
+
+* Add simple in-browser visualizations.
+  > These visualizations can be served via backend using Python library like
+  > Bokeh or rendered on the client using D3 (or equivalent) library. I prefer the
+  > latter approach.
+
+### Testing
+
+The client-side needs to work reliably across different systems, mainly Linux,
+Windows and OSX machines.
+
+By default, I will develop and test the scripts on my Linux system. For Windows
+and OSX, however, I will have to take help of mentors/community members or my
+friends.
+
+### Updating installation scripts
+
+Modify the installation-test scripts to generate a diagnostic report containing
+machine information and status report as described before. The script should:
+
+* catch abrupt halts (`SIGINT`, for example) and report it to the server.
```

Comment:
> I'm against reporting automatically
Okay. Probably an exit message like, 'Script failed. Please run with -foo switch to report any messages to server.' would be good.

> how would that be saved in your proposed schema?

We'll be sending the data to the server through POST via JSON. We can directly save this file as a log.